cpp/src/arrow/compute/exec/asof_join_node.cc

```diff
@@ -657,34 +986,85 @@ class AsofJoinNode : public ExecNode {
     return std::make_shared<arrow::Schema>(fields);
   }
 
+  static inline Result<col_index_t> FindColIndex(const Schema& schema,
+                                                 const FieldRef& field_ref,
+                                                 const util::string_view& key_kind) {
```

Comment:
Why do we use string_view here instead of just string? (I don't know what the difference is )

Reply:
AFAIK, the difference is that a literal string, such as `"on"` or `"by"` that [the callers](https://github.com/apache/arrow/pull/13880/files#diff-5789a42aebc3bd0f5a6db687c78ab0c3eef3e316982f99fe9e0ea21fabed354cR1024-R1028) pass, will get wrapped with zero-copy by `string_view` instead of copied into a `string`.

Reply:
`string_view` is roughly

```
{
  char* str;
  size_t len;
}
```

So it's basically "pointer + size".  You are correct that we can zero-copy from `string` into `string_view`.  However, the `to_string` you call below is not zero-copy and you would probably be better off with `const std::string& key_kind`.

Also, in the future, when you do want to use `string_view`, you should just do `string_view key_kind` and not `const string_view& key_kind` as it is small enough it doesn't need to be passed by reference and is better passed by value (like a pointer).

Reply:
You're right, however `to_string` is only invoked upon error, which can be assumed to be infrequent. I'll fix to passing by-value.